### PR TITLE
Cleanup fmod vs IEEE remainder

### PR DESCRIPTION
**Pull request by tautschnig**
1. Add tests for both the C and Java front-end.
2. Make sure we support both variants of the modulus operator on floating-point values in the back-ends: the SMT back-end only supported remainder (and only for FPA mode), the propositional back-end did not distinguish between fmod and IEEE remainder.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
Asking for help from @martin-cs: the remaining issue is that we need an ability to round-to-integer (`roundToIntegral` in SMT2) in our back-ends.

**Comment by kroening** [comment]
May I suggest pulling out the FMA? It's independently useful, e.g., for systolic arrays.

**Comment by tautschnig** [comment]
> May I suggest pulling out the FMA? It's independently useful, e.g., for systolic arrays.

Will do, but I have a few more updates that are still being completed. Will create a separate PR once those are done.

File: src/util/floatbv_expr.h [hunk 1]
```diff
@@ -568,6 +568,71 @@ inline floatbv_rem_exprt &to_floatbv_rem_expr(exprt &expr)
   return static_cast<floatbv_rem_exprt &>(expr);
 }
 
+/// \brief Fused multiply-add expression: round(op0 * op1 + op2)
+/// with a single rounding. No explicit rounding mode operand;
+/// the solver uses the ambient rounding mode.
+class floatbv_fma_exprt : public multi_ary_exprt
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply src/util/floatbv_expr.h]: Fixed. The comment now correctly states the rounding mode is stored as the 4th operand.

File: src/solvers/flattening/boolbv_floatbv_mod_rem.cpp [hunk 2]
```diff
@@ -19,7 +19,10 @@ bvt boolbvt::convert_floatbv_mod_rem(const binary_exprt &expr)
 
   float_utilst float_utils(prop);
 
-  auto rm = bv_utils.build_constant(ieee_floatt::ROUND_TO_EVEN, 32);
+  auto rm = bv_utils.build_constant(
+    expr.id() == ID_floatbv_rem ? ieee_floatt::ROUND_TO_EVEN
+                                : ieee_floatt::ROUND_TO_ZERO,
+    32);
   float_utils.set_rounding_mode(rm);
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply src/solvers/flattening/boolbv_floatbv_mod_rem.cpp:26]: Updated the comment to reflect the current implementation.

File: src/solvers/floatbv/float_bv.cpp [hunk 3]
```diff
@@ -132,6 +132,22 @@ exprt float_bvt::convert(const exprt &expr) const
...
+  {
+    const auto &float_expr = to_binary_expr(expr);
+    return rem(float_expr.lhs(), float_expr.rhs());
+  }
+  else if(expr.id() == ID_floatbv_fma)
+  {
+    // float_bvt does not implement FMA; fall through to default handling
+    // which will report an error. A proper implementation would require
+    // double-width multiplication followed by addition and single rounding.
+  }
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply src/solvers/floatbv/float_bv.cpp:145]: Implemented proper single-rounding FMA in `float_bvt`: multiply at double width (exact), align and add the third operand (exact), round once. This matches `float_utilst::fma` and ensures correctness of the remainder algorithm.

File: src/solvers/floatbv/float_utils.cpp [hunk 4]
```diff
@@ -575,20 +684,106 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+  bvt exp_diff = bv_utils.sub(exp1, exp2);
+  literalt ex_ge_ey = !exp_diff.back();
+  bvt abs_exp_diff = bv_utils.absolute_value(exp_diff);
+
+  // Integer width for aligned significands
+  const std::size_t int_width = (std::size_t(1) << spec.e) + frac_bits + 2;
+  bvt shift_dist = limit_distance(abs_exp_diff, mp_integer(int_width));
+
+  bvt mx = bv_utils.zero_extension(unpacked1.fraction, int_width);
+  bvt my = bv_utils.zero_extension(unpacked2.fraction, int_width);
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp]: Added a comment documenting that the width is O(2^e) bits, feasible for half/float/double but infeasible for long double/quad. The SMT FPA backend handles large formats efficiently.

File: src/solvers/floatbv/float_utils.cpp [hunk 5]
```diff
@@ -575,20 +684,106 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+  fmod_unpacked.infinity = const_literal(false);
+  fmod_unpacked.zero = bv_utils.is_zero(r_int);
+  bvt fmod_result = round_and_pack(fmod_unpacked);
+
+  // Handle special cases
+  literalt x_is_special =
+    prop.lor({unpacked1.zero, unpacked1.infinity, unpacked1.NaN});
+  fmod_result = bv_utils.select(x_is_special, src1, fmod_result);
+  fmod_result = bv_utils.select(unpacked2.infinity, src1, fmod_result);
+
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp]: Fixed. `fmod(±inf, y)`, `fmod(x, ±0)`, `fmod(x, NaN)`, and `fmod(NaN, y)` now correctly return NaN.

**Comment by tautschnig** [comment]
> > May I suggest pulling out the FMA? It's independently useful, e.g., for systolic arrays.
> 
> Will do, but I have a few more updates that are still being completed. Will create a separate PR once those are done.

Now in https://github.com/diffblue/cbmc/pull/8884. The present PR should eventually be rebased on top of that.

**Comment by tautschnig** [comment]
> Now in #8884. The present PR should eventually be rebased on top of that.

Rebase now done.

File: regression/cbmc-library/fmod/main.c [hunk 6]
```diff
@@ -1,9 +1,19 @@
...
 int main()
 {
-  fmod();
-  assert(0);
-  return 0;
+  // examples from
+  // https://stackoverflow.com/questions/2947044/how-do-i-use-modulus-for-float-double
+  // and
+  // https://stackoverflow.com/questions/25734144/difference-between-c-functions-remainder-and-fmod
+  double d1 = fmod(0.5, 0.3);
```
Comment (martin-cs) [inline regression/cbmc-library/fmod/main.c:15]: '0.3' as a decimal makes me nervous.

File: regression/cbmc-library/remainderf/_Float16.c [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+#if defined(__GNUC__) && __GNUC__ >= 13
+
+_Float16 __CPROVER_remainderf16(_Float16, _Float16);
+
+int main()
+{
+  _Float16 pos_inf_f16 = (_Float16)(1.0 / 0.0);
+  _Float16 x, y;
+  __CPROVER_assume(y != (_Float16)0.0);
```
Comment (martin-cs) [inline regression/cbmc-library/remainderf/_Float16.c:36]: Why ignore the NaN cases?

File: src/solvers/floatbv/float_bv.cpp [hunk 8]
```diff
@@ -953,6 +963,87 @@ exprt float_bvt::fma(
...
+{
+  PRECONDITION(x.type() == y.type());
+  const floatbv_typet &type = to_floatbv_type(x.type());
+  const ieee_float_spect spec{type};
+
+  // x - round-to-integer-towards-zero(x / y) * y
+  const constant_exprt round_to_zero =
+    from_integer(ieee_floatt::ROUND_TO_ZERO, unsignedbv_typet{2});
+
+  exprt div = convert(ieee_float_op_exprt{x, ID_floatbv_div, y, round_to_zero});
```
Comment (martin-cs) [inline src/solvers/floatbv/float_bv.cpp:980]: This feels like a different semantics to the one below.

File: src/solvers/floatbv/float_utils.cpp [hunk 9]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+  // Exponent difference
+  bvt exp1 =
+    bv_utils.sign_extension(unpacked1.exponent, unpacked1.exponent.size() + 1);
+  bvt exp2 =
+    bv_utils.sign_extension(unpacked2.exponent, unpacked2.exponent.size() + 1);
+  bvt exp_diff = bv_utils.sub(exp1, exp2);
+  literalt ex_ge_ey = !exp_diff.back();
+  bvt abs_exp_diff = bv_utils.absolute_value(exp_diff);
+
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp:735]: Exponent difference seems correct and is important.

File: src/solvers/floatbv/float_utils.cpp [hunk 10]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+  // Integer width for aligned significands.
+  // Note: this is O(2^e) bits, which is feasible for half (45 bits),
+  // float (282 bits), and double (2099 bits), but infeasible for
+  // long double/quad (32834 bits). Use the SMT FPA backend for those.
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp]: I'm not convinced there are any SMT solvers that will handle this better!  (Yet...)

Essentially this is the expensive bit of all approaches; you may need to do "exponent-difference" divide steps.

File: src/solvers/floatbv/float_utils.cpp [hunk 11]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+  const std::size_t int_width = (std::size_t(1) << spec.e) + frac_bits + 2;
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp]: I'm not sure this is quite right.  For example for single precision you have 127 as the maximum exponent, -149 as the smallest and so the maximum difference is 276.  So if you want to avoid overflowing then you need 276+24 = 300 bits.  Maybe I am doing my calculations wrong.  Note that these cases are rare and generally simple in other ways.

File: src/solvers/floatbv/float_utils.cpp [hunk 12]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+  bvt shift_dist = limit_distance(abs_exp_diff, mp_integer(int_width));
+
+  bvt mx = bv_utils.zero_extension(unpacked1.fraction, int_width);
+  bvt my = bv_utils.zero_extension(unpacked2.fraction, int_width);
+
+  // Align: shift the one with larger exponent left
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp:751]: Yes, hard agree!

File: src/solvers/floatbv/float_utils.cpp [hunk 13]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+    ex_ge_ey,
+    bv_utils.shift(mx, bv_utilst::shiftt::SHIFT_LEFT, shift_dist),
+    mx);
+  bvt my_aligned = bv_utils.select(
+    ex_ge_ey,
+    my,
+    bv_utils.shift(my, bv_utilst::shiftt::SHIFT_LEFT, shift_dist));
+
+  // Integer remainder: fmod significand (unsigned)
+  bvt r_int = bv_utils.remainder(
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp:762]: Yes.  I would imagine this is most of the cost.

File: src/solvers/floatbv/float_utils.cpp [hunk 14]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+    mx_aligned, my_aligned, bv_utilst::representationt::UNSIGNED);
+
+  // Integer quotient LSB (needed for remainder tie-breaking)
+  bvt q_int = bv_utils.divider(
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp:768]: IIRC you only need the LSB of the divider or something like that and you can do it by tapping the penultimate round of the remainder.  There might be something like doing one fewer divide step rounds and then doing the last one explicitly.

File: src/solvers/floatbv/float_utils.cpp [hunk 15]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+    bv_utils.build_constant(
+      mp_integer(int_width) - mp_integer(frac_bits), spec.e + 2));
+  unbiased_floatt fmod_unpacked;
+  fmod_unpacked.fraction = r_int;
+  fmod_unpacked.exponent = adjusted_exp;
+  fmod_unpacked.sign = unpacked1.sign;
+  fmod_unpacked.NaN = const_literal(false);
+  fmod_unpacked.infinity = const_literal(false);
+  fmod_unpacked.zero = bv_utils.is_zero(r_int);
+  bvt fmod_result = round_and_pack(fmod_unpacked);
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp:794]: The result should be exact so you shouldn't need to round.  I think with the architecture of the CProver encodings you may not have any choice though.

File: src/solvers/floatbv/float_utils.cpp [hunk 16]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+
+  // Handle IEEE 754 special cases:
+  //   fmod(x, ±0)    = NaN
+  //   fmod(±inf, y)   = NaN
+  //   fmod(NaN, y)    = NaN
+  //   fmod(x, NaN)    = NaN
+  //   fmod(±0, y)     = ±0 (= x)
+  //   fmod(x, ±inf)   = x
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp:804]: From memory I think these are correct.  They should be listed in SymFPU thought.

File: src/solvers/floatbv/float_utils.cpp [hunk 17]
```diff
@@ -690,20 +690,142 @@ bvt float_utilst::div(const bvt &src1, const bvt &src2)
...
+  // y is ±inf and no NaN condition → return x
+  fmod_result = bv_utils.select(
+    prop.land(unpacked2.infinity, !nan_result), src1, fmod_result);
+
+  // For fmod (ROUND_TO_ZERO), we're done
+  bvt result = fmod_result;
+
+  if(!rounding_mode_bits.round_to_zero.is_true())
+  {
+    // Step 2: remainder(fmod, y) via FMA. |fmod/y| < 1, n ∈ {-1,0,1}.
```
Comment (martin-cs) [inline src/solvers/floatbv/float_utils.cpp]: I am not sure I see why this is needed.  I think you can get different rounding directions with a simple conditional subtract of y, ideally before you pack.

File: src/solvers/floatbv/float_utils.cpp (hunk 9 above)
Comment (martin-cs) [reply src/solvers/floatbv/float_utils.cpp:735]: Oh, also, you can use the sign bit to fast path because x < y => rem(x,y) = x in many cases.

**Review by martin-cs — COMMENTED** [review]
I'm going to have to think about it a bit...

File: regression/cbmc-library/fmod/main.c (hunk 6 above)
Comment (tautschnig) [reply regression/cbmc-library/fmod/main.c:15]: Added comment explaining `0.3`/`0.2` are not exactly representable but results are correct on
nearest representable values. Also added hex float literal variants (`0x1.0p-1`, `0x1.3333333333333p-2`, `0x1.999999999999ap-3`).

File: regression/cbmc-library/remainderf/_Float16.c (hunk 7 above)
Comment (tautschnig) [reply regression/cbmc-library/remainderf/_Float16.c:36]: Added NaN edge case tests: `remainder(NaN, y)`, `remainder(x, NaN)`, `remainder(x, 0)`, `remainder(inf, y)` - all assert the result is `NaN`. Added comment explaining the finite-input focus of the property test.

File: src/solvers/floatbv/float_bv.cpp (hunk 8 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_bv.cpp:980]: Added comment clarifying fmod uses truncation toward zero (C99 §7.12.10.1) while rem uses round-to-nearest-even (C99 §7.12.10.2).

File: src/solvers/floatbv/float_utils.cpp (hunk 9 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp:735]: Thanks, implemented `|x| < |y| -> result = x` fast path. Added at the beginning (detection) and end (selection) of the function.

File: src/solvers/floatbv/float_utils.cpp (hunk 11 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp]: Yes, that's why this shouldn't be built without your oversight! Fixed to `(1 << e) + 2*p - 2`, which correctly accounts for `max_exp_diff = 2^e + p - 4` plus `p` significand bits. Updated comment with derivation.

File: src/solvers/floatbv/float_utils.cpp (hunk 15 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp:794]: Added comment explaining the fmod result is mathematically exact but `round_and_pack` is needed for normalization (subnormals, exponent adjustment).

File: src/solvers/floatbv/float_utils.cpp (hunk 16 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp:804]: Reference added.

File: src/solvers/floatbv/float_utils.cpp (hunk 14 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp:768]: Ack, but left this as a comment rather than implementing this for now.

Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp:768]: IMHO, we should independently look at the divider implementation and see whether that could be improved.

File: src/solvers/floatbv/float_utils.cpp (hunk 9 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp:735]: Reverted as `|x| < |y|` does not imply `remainder(x,y) = x` when `|x| > |y|/2`. Concrete example: `remainder(0.7, 1.0)`. `x/y = 0.7`, nearest integer is `1`, so `remainder = 0.7 - 1.0 = -0.3`. The fast path would have incorrectly returned `0.7`.

File: src/solvers/floatbv/float_utils.cpp (hunk 17 above)
Comment (tautschnig) [reply src/solvers/floatbv/float_utils.cpp]: Thank you, this makes a huge difference - see [3580205](https://github.com/diffblue/cbmc/pull/7885/commits/35802057a98556068da80ea19c2772062dca1fe9), including updated performance numbers.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 18]
```diff
@@ -1659,9 +1659,9 @@ static smt_termt dispatch_expr_to_smt_conversion(
     return convert_expr_to_smt(*multiply, converted);
   }
 #if 0
-  else if(expr.id() == ID_floatbv_rem)
```
Comment (kroening) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:1662]: This change doesn't do anything.

File: regression/smt2_solver/fp-rem-nonstandard/fp-rem1.desc [hunk 19]
```diff
@@ -0,0 +1,9 @@
+KNOWNBUG
+fp-rem1.smt2
+
+^EXIT=0$
+^SIGNAL=0$
+^unsat$
+--
+Based on Z3#2381: fp.rem(3.0, 2.0) should be -1.0.
+CBMC's fp.rem appears to always return +0.0 (bug).
```
Comment (kroening) [inline regression/smt2_solver/fp-rem-nonstandard/fp-rem1.desc:9]: Is this now fixed?

File: src/solvers/smt2/smt2_conv.h [hunk 20]
```diff
@@ -146,6 +146,7 @@ class smt2_convt : public stack_decision_proceduret
   void convert_floatbv_minus(const ieee_float_op_exprt &expr);
   void convert_floatbv_div(const ieee_float_op_exprt &expr);
   void convert_floatbv_mult(const ieee_float_op_exprt &expr);
+  void convert_floatbv_mod(const binary_exprt &expr);
```
Comment (kroening) [inline src/solvers/smt2/smt2_conv.h:149]: floatbv_mod_exprt

File: src/solvers/smt2/smt2_conv.h [hunk 21]
```diff
@@ -146,6 +146,7 @@ class smt2_convt : public stack_decision_proceduret
...
   void convert_floatbv_rem(const binary_exprt &expr);
```
Comment (kroening) [inline src/solvers/smt2/smt2_conv.h:150]: floatbv_rem_exprt

Comment (tautschnig) [reply src/solvers/smt2/smt2_conv.h]: Done.

File: src/solvers/smt2/smt2_conv.h (hunk 20 above)
Comment (tautschnig) [reply src/solvers/smt2/smt2_conv.h]: Done.

File: regression/smt2_solver/fp-rem-nonstandard/fp-rem1.desc (hunk 19 above)
Comment (tautschnig) [reply regression/smt2_solver/fp-rem-nonstandard/fp-rem1.desc]: Cleaned up.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 18 above)
Comment (tautschnig) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Cleaned and will be moved to a PR of its own.
